internal/net/net.go

```diff
@@ -98,16 +116,22 @@ func SplitHostPort(hostport string) (host string, port uint16, err error) {
 	return host, port, err
 }
 
+// ScanPorts scans the given range of port numbers from the host (inclusively),
+// and return the list of ports that can be connected through TCP, or any error occurred.
 func ScanPorts(ctx context.Context, start, end uint16, host string) (ports []uint16, err error) {
+	if start > end {
+		start, end = end, start
+	}
+
 	var rl syscall.Rlimit
-	err = syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rl)
-	if err != nil {
+	if err = syscall.Getrlimit(syscall.RLIMIT_NOFILE, &rl); err != nil {
 		return nil, err
 	}
 	eg, egctx := errgroup.New(ctx)
 	eg.Limitation(int(rl.Max) / 2)
+
 	var mu sync.Mutex
-	for i := start; i <= end; i++ {
+	for i := start; i >= start && i <= end; i++ {
```

Comment:
Could you show me the reason?

Reply:
The valid port number is from 1 to 65535.
it is because when `i = 65534` and execute `i++`, which add i to 65535, the uint16 overflow and reset `i` to 0, causing infinity loop. 
This checking checks if the `i` overflow and prevent the infinity loop.

Reply:
It's so complicated problem, the variable `start` and `end` is uint16 variable so `i` becomes uint16.
 the case of `i == math.MaxUint16` then `i` will continue to next incremental loop due to address overflow `i` becomes 0 and continue to loop end

Reply:
you can reproduce it here
https://play.golang.org/p/Z-Nvjo7Ewu0